datastream-common/src/main/java/com/linkedin/datastream/metrics/BrooklinMetricInfo.java

```diff
@@ -32,4 +35,34 @@ protected BrooklinMetricInfo(String nameOrRegex, Optional<List<String>> attribut
   public String getNameOrRegex() {
     return _nameOrRegex;
   }
+
+  private boolean equalAttributes(Optional<List<String>> a, Optional<List<String>> b) {
+    if (!a.isPresent() || !b.isPresent()) {
+      return a.isPresent() == b.isPresent();
+    }
+
+    // Both are present
+    return Objects.equals(new HashSet<>(a.get()), new HashSet<>(b.get()));
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    BrooklinMetricInfo that = (BrooklinMetricInfo) o;
+    return _nameOrRegex.equals(that._nameOrRegex) && equalAttributes(_attributes, that._attributes);
+  }
+
+  @Override
+  public int hashCode() {
+    if (_attributes.isPresent()) {
+      return Objects.hash(_nameOrRegex, new HashSet<>(_attributes.get()));
+    } else {
+      return Objects.hash(_nameOrRegex);
+    }
```

Comment:
Since this class is immutable, I suggest we precompute and cache the value this function returns in a private final field upon construction.